java/ql/src/Stubs/Stubs.qll

```diff
@@ -0,0 +1,440 @@
+  exists(Type t, string suff | result = stubTypeName(t) + suff + " " + p.getName() |
+    if p.isVarargs()
+    then (
+      t = p.getType().(Array).getElementType() and
+      suff = "..."
+    ) else (
+      t = p.getType() and suff = ""
+    )
+  )
+}
+
+private string stubEnumConstants(RefType t) {
+  if t instanceof EnumType
+  then
+    exists(EnumType et | et = t |
+      result =
+        "    " + concat(EnumConstant c | c = et.getAnEnumConstant() | c.getName(), ", ") + ";\n"
+    )
+  else result = ""
+}
+
+// Holds if the member is to be excluded from stubMember
+private predicate excludedMember(Member m) {
+  m instanceof EnumConstant
+  or
+  exists(Method c | m = c |
+    c.getDeclaringType() instanceof EnumType and
+    m.hasName(["values", "valueOf"]) and
+    m.isStatic()
+  )
+}
+
+private string stubMember(Member m) {
+  if excludedMember(m)
+  then result = ""
+  else (
+    exists(Method c | m = c |
+      result =
+        "    " + stubModifiers(c) + stubGenericMethodParams(c) + stubTypeName(c.getReturnType()) +
+          " " + c.getName() + "(" + stubParameters(c) + ")" + stubImplementation(c) + "\n"
+    )
+    or
+    exists(Constructor c | m = c |
+      result =
+        "    " + stubModifiers(m) + c.getName() + "(" + stubParameters(c) + ")" +
+          stubImplementation(c) + "\n"
+    )
+    or
+    exists(Field f | f = m |
+      result =
+        "    " + stubModifiers(m) + stubTypeName(f.getType()) + " " + f.getName() + " = " +
+          stubDefaultValue(f.getType()) + ";\n"
+    )
+    or
+    exists(NestedType nt | nt = m | result = indent(nt.(GeneratedType).getStub()))
+  )
+}
+
+bindingset[s]
+private string indent(string s) { result = "    " + s.replaceAll("\n", "\n    ") + "\n" }
+
+// If a class's superclass doesn't have a no-arg constructor, then it won't compile when its constructor's bodies are stubbed
+// So we synthesise no-arg constructors for each generated type that doesn't have one.
+private string stubFakeConstructor(RefType t) {
+  if not t instanceof Class
+  then result = ""
+  else
+    exists(string mod |
+      if t instanceof EnumType then mod = "    private " else mod = "    protected "
+    |
+      if hasNoArgConstructor(t) then result = "" else result = mod + t.getName() + "() {}\n"
+    )
+}
+
+private predicate hasNoArgConstructor(Class t) {
+  exists(Constructor c | c.getDeclaringType() = t |
+    c.getNumberOfParameters() = 0 and
+    not c.isPrivate()
+  )
+}
+
+private RefType getAReferencedType(RefType t) {
+  result = t.(GeneratedType).getAGeneratedType()
+  or
+  result =
+    getAReferencedType(any(NestedType nt |
+        nt.getEnclosingType().getSourceDeclaration() = t.getSourceDeclaration()
+      ))
+  or
+  exists(RefType t1 | t1 = getAReferencedType(t) |
+    result = t1.(NestedType).getEnclosingType()
+    or
+    result = t1.getSourceDeclaration()
+    or
+    result = t1.(ParameterizedType).getATypeArgument()
+    or
+    result = t1.(Array).getElementType()
```

Comment:
Does this recursion handle `A[][]` or `Map<Map<A, ...>, ...>`?

Reply:
It should do, yes.